lib/ecto/schema.ex

```diff
@@ -754,6 +780,12 @@ defmodule Ecto.Schema do
     Ecto.Model.Callbacks.__apply__(struct.__struct__, :after_load, loaded)
   end
 
+  def __load__(struct, source, fields, map, id_types) do
```

Comment:
I have an idea. What if instead of calling `__load__` from the mongo adapter, you simply return a map and we add a clause here that will match the `{:&, _, _}` against the map and call `__load__`? https://github.com/elixir-lang/ecto/blob/master/lib/ecto/repo/queryable.ex#L143

Since we will start handling maps and embeds, I guess we should handle this case automatically too. Keep in mind though the SQL adapters will already return a struct, in such cases, `__load__` should be a no-op (i.e. if a struct is given).